### ISSUE TITLE
Spike: Explore and investigate ongoing version upgrade for HL Fabric network  

### ISSUE THREAD

**Issue by suvajit-sarkar** [hyperledger/bevel#1765]
### Description
As a developer I want to explore and investigate if it is possible for ongoing version upgrade of HL Fabric network.
For example upgrade from v1.4.8 to v2.2.2 

### Acceptance Criteria
1. Outcomes documented and further issues created

**Comment by TonyRowntree** [comment]
I am working on this

**Comment by TonyRowntree** [comment]
https://jira.hyperledger.org/browse/FAB-16986, https://hyperledger-fabric.readthedocs.io/en/release-2.2/upgrade.html, https://hyperledger-fabric.readthedocs.io/en/release-2.2/upgrade_to_newest_version.html

**Comment by TonyRowntree** [comment]
Find version that is stable > 2.2.2

**Comment by TonyRowntree** [comment]
Report: 

Each fabric v2.x minor release builds on the v2.0 release with minor features, improvements, and bug fixes.

Fabric v2.2 is the first long-term support (LTS) release of Fabric v2.x. Fixes will be provided on the v2.2.x release stream until after the next LTS release is announced.

As Hyperledger Bevel is using Fabric 2.2.0 upgrades to 2.x will be possible with no to minor disruption.

https://hyperledger-fabric.readthedocs.io/en/release-2.2/whatsnew.html

**Comment by TonyRowntree** [comment]
Upgrade from LST 1.4 to LTS 2.2 is possible although there are a few points to consider:

1) Chaincode lifecycle - The new chaincode lifecycle allows multiple organizations to agree on how a chaincode will be operated before it can be used on a channel. It is a best practice to upgrade all of the peers on a channel before enabling the Channel and Application capabilities that enable the new chaincode lifecycle. Any peers that are not at v2.x will crash after enabling either capability, any ordering nodes that are not at v2.x will crash after the Channel capability has been enabled. [1]

2) Chaincode shim changes (Go chaincode only) - The v2.x ccenv image that is used to build Go chaincodes no longer automatically vendors the Go chaincode shim dependency like the v1.4 ccenv image did. The recommended approach is to vendor the shim in your v1.4 Go chaincode before making upgrades to the peers and channels, since this approach works with both a v1.4.x and v2.x peer. [2]

3) Peer databases upgrade - During the process for upgrading your peers, you will need to perform one additional step to upgrade the peer databases. The databases of all peers (which include not just the state database but the history database and other internal databases for the peer) must be rebuilt using the v2.x data format as part of the upgrade to v2.x. To trigger the rebuild, the databases must be dropped before the peer is started. [3]

In summary, the upgrade from 1.4LTS to 2.2 LTS is possible as long as all the steps are carefully followed.

source: [1] [2] [3] - hyperledger-fabric.readthedocs.io/en/release-2.2/upgrade_to_newest_version.html

**Comment by jagpreetsinghsasan** [comment]
> Upgrade Fabric from 1.4x to 2.2x is possible although there are a few points to consider:
> 
> 1. Chaincode lifecycle - The new chaincode lifecycle allows multiple organizations to agree on how a chaincode will be operated before it can be used on a channel. It is a best practice to upgrade all of the peers on a channel before enabling the Channel and Application capabilities that enable the new chaincode lifecycle. Any peers that are not at v2.x will crash after enabling either capability, any ordering nodes that are not at v2.x will crash after the Channel capability has been enabled. [1]
> 2. Chaincode shim changes (Go chaincode only) - The v2.x ccenv image that is used to build Go chaincodes no longer automatically vendors the Go chaincode shim dependency like the v1.4 ccenv image did. The recommended approach is to vendor the shim in your v1.4 Go chaincode before making upgrades to the peers and channels, since this approach works with both a v1.4.x and v2.x peer. [2]
> 3. Peer databases upgrade - During the process for upgrading your peers, you will need to perform one additional step to upgrade the peer databases. The databases of all peers (which include not just the state database but the history database and other internal databases for the peer) must be rebuilt using the v2.x data format as part of the upgrade to v2.x. To trigger the rebuild, the databases must be dropped before the peer is started. [3]
> 
> In summary, the upgrade from 1.4x to 2.2x is possible as long as all the steps are carefully followed.
> 
> source: [1] [2] [3] - hyperledger-fabric.readthedocs.io/en/release-2.2/upgrade_to_newest_version.html

There will be much more changes as we have to upgrade the entities.
I have created a [spike](https://github.com/hyperledger/bevel/issues/1845) listing the actual process, in accordance to our charts and flow

**Comment by sownak** [comment]
Closing this as now this will be tracked as #1845
